users/luescher/setups/hybrid/gmm_system.py

```diff
@@ -0,0 +1,845 @@
+            name=name,
+            corpus=self.corpora[name],
+            concurrent=self.concurrent[name],
+            segment_path=segm_corpus_job.segment_path,
+        )
+        self.jobs[name]["segment_corpus"] = segm_corpus_job
+
+    @tk.block()
+    def _init_lm(self, name, file, type, scale, **kwargs):
+        self.crp[name].language_model_config = rasr.RasrConfig()
+        self.crp[name].language_model_config.type = type
+        self.crp[name].language_model_config.file = file
+        self.crp[name].language_model_config.scale = scale
+
+    @tk.block()
+    def _init_lexica(self, name, file, normalize_pronunciation, **kwargs):
+        self.crp[name].lexicon_config = rasr.RasrConfig()
+        self.crp[name].lexicon_config.file = file
+        self.crp[name].lexicon_config.normalize_pronunciation = normalize_pronunciation
+
+    def allow_zero_weights(self, name: str = "base"):
+        if self.crp[name].acoustic_model_post_config is None:
+            self.crp[name].acoustic_model_post_config = rasr.RasrConfig()
+        self.crp[name].acoustic_model_post_config.mixture_set.allow_zero_weights = True
+        self.crp[
+            name
+        ].acoustic_model_post_config.old_mixture_set.allow_zero_weights = True
+
+    def _add_features(self, name, corpus, feature_job, prefix=""):
+        self.jobs[corpus][name] = feature_job
+        feature_job.add_alias(f"{prefix}{corpus}_{name}_features")
+        self.feature_caches[corpus][name] = feature_job.out_feature_path[name]
+        self.feature_bundles[corpus][name] = feature_job.out_feature_bundle[name]
+        feature_path = rasr.FlagDependentFlowAttribute(
+            "cache_mode",
+            {
+                "task_dependent": self.feature_caches[corpus][name],
+                "bundle": self.feature_bundles[corpus][name],
+            },
+        )
+        self.feature_flows[corpus][name] = features.basic_cache_flow(feature_path)
+        self.feature_flows[corpus][f"uncached_{name}"] = feature_job.feature_flow
+
+    @tk.block()
+    def extract_features(self, feat_args: dict, **kwargs):
+        corpus_list = self.train_corpora + self.dev_corpora + self.test_corpora
+
+        if "mfcc" in feat_args.keys():
+            for c in corpus_list:
+                self.mfcc_features(c, **feat_args["mfcc"])
+                self.energy_features(c, **feat_args["energy"])
+            for t in self.train_corpora:
+                self.add_energy_to_features(t, "mfcc+deriv")
+
+        if "gt" in feat_args.keys():
+            for c in corpus_list:
+                self.gt_features(c, **feat_args["gt"])
+                self.energy_features(c, **feat_args["energy"])
+            for t in self.train_corpora:
+                self.add_energy_to_features(t, "gt")
+
+        if "fb" in feat_args.keys():
+            for c in corpus_list:
+                self.fb_features(c, **feat_args["fb"])
+
+        unknown_features = set(feat_args.keys()).difference(
```

Comment:
Maybe we can allow for a generic feature type when the extraction flow is fully given. (E.g. NN-Bottleneck features)

Reply:
so just add "generic"?
wouldn't the `extract_features` function just be extended by the user? or how do you want to handle this? Have a call to the FeatureExtractionJob?

Reply:
I would instead of throwing an error if `unknown_features` are non-empty, loop over these and then call a `self.generic_features(c, feature_name, **feat_args[feature_name])` function (we can expand meta.system) that creates a generic feature extraction job.

the MfccJob is only a convenience function anyway.

Reply:
I changed the pipeline for the feature extraction. It now allows also for generic feature extraction.